### DISCUSSION TITLE
How to move AuthorizationState from AuthorizationStateWaitPassword back to AuthorizationStateWaitOtherDeviceConfirmation?

### DISCUSSION THREAD

**Discussion by yvasyliev** [tdlight-team/tdlight-java#194] [Q&A]
Let's say the user decided not to enter the password after QR code scan. 

How to cancel current `AuthorizationStateWaitPassword` state?

If I restart the app, I'll get `AuthorizationStateWaitPassword` state right from the startup.

I can write some kind of hook that deletes session directory, and then I'll get `AuthorizationStateWaitOtherDeviceConfirmation` state, but is there a better variant?

**Comment by cavallium** [comment]
Does it work if you send TdApi.LogOut() instead?

**Reply by yvasyliev** [reply]
When `client.send(new TdApi.LogOut())` is done, I'm receiving `AuthorizationStateClosed`. In this state `SimpleTelegramClient` becomes unusable. And if I restart the app, I'll get `AuthorizationStateWaitOtherDeviceConfirmation`

**Reply by yvasyliev** [reply]
btw need to check what will be the behavior if I create new `SimpleTelegramClient` from a `SimpleTelegramClientBuilder` after I call `TdApi.LogOut` on previous one

**Reply by yvasyliev** [reply]
@cavallium I might need some help here.

I switched to `it.tdlight.TelegramClient`.

The flow is the following:
1. Create `ClientFactory clientFactory`
2. Create `TelegramClient telegramClient`
3. Initialize `telegramClient`
4. Receive `TdApi.AuthorizationStateWaitPhoneNumber` update
5. Call `TdApi.RequestQrCodeAuthentication` function
6. Receive `TdApi.AuthorizationStateWaitOtherDeviceConfirmation` update
7. Scan QR code
8. Receive `TdApi.AuthorizationStateWaitPassword` update
9. Call `TdApi.LogOut`
10. Receive `TdApi.AuthorizationStateLoggingOut` update
11. Receive `TdApi.AuthorizationStateClosed` update
12. `telegramClient = clientFactory.createClient()`
13. Initialize `telegramClient`

It looks like after `TdApi.AuthorizationStateClosed` it is not possible to reuse `clientFactory.createClient()`. 
Because if I try to perform any action with re-initialized `telegramClient`, I'll get the following error:
```
Error {
  code = 503
  message = "Client closed"
}
```

Could you please confirm?

I can share full project example later, if you need.

**Reply by cavallium** [reply]
the clientFactory can be used with multiple clients, it's strange that it does not allow you to re-create a client.
Yes, share the project, thanks.

**Reply by yvasyliev** [reply]
Here is my project: https://github.com/yvasyliev/telegram-dynamic-profile-photo/tree/soring-boot-migration

It's a Spring Boot projects that runs Telegram client and Telegram bot in parallel. All client interactions go through bot.

To reproduce my issue:
1. Run the app with the following propeties:
    ```properties
    telegram.api.hash=${API_HASH}
    telegram.api.id=${API_ID}
    telegram.bot.token=${BOT_TOKEN}
    telegram.bot.username=${BOT_USERNAME}
    ```
    - set them as environment virables
    - or add them to `application.properties`
2. Send `/tglogin` command to bot
3. Scan QR code
4. Send `/cancel` to bot

After these actions re-initialized `TelegramClient` gets unresponsive.

Please ignore `README.md`. 😊

**Reply by cavallium** [reply]
Hi, unfortunately I never tried to change/cancel an authentication method, maybe you should ask in TDLib chat to Levlam: https://t.me/tdlibchat (tdlib API is identical to tdlight).
Ask which command must be sent to cancel a login, and how to handle the related updates.

**Reply by yvasyliev** [reply]
I've got the same suggestion from `tdlibchat` as yours - call `TdApi.LogOut`. 🙂

If I create new `TelegramClient` instance, I'll get this error:
```
Error {
  code = 503
  message = "Client closed"
}
```

It means that I'm still using old TdLib instance, although `clientFactory.createClient()` has been invoked second time.
Also, the new `TelegramClient` doesn't receive any updates, so I guess there's something wrong with `ClientFactory`.

I tried to create new instance of `ClientFactory` every time I get `AuthorizationStateClosed`, and I got progress!
<details>
    <summary>Take a look at console logs:</summary>

    
    11:17:12.739 [tgphotomanagerbot Telegram Executor] TRACE it.tdlight.TelegramClient - Trying to send async request LogOut {
    }

    11:17:12.739 [TDLib thread] TRACE it.tdlight.ResponseReceiver - Received 2 events
    11:17:12.739 [TDLib thread] TRACE it.tdlight.TelegramClient - Client 1, response received for request 0: UpdateAuthorizationState {
    authorizationState = AuthorizationStateLoggingOut {
    }
    }

    11:17:12.740 [TDLib thread] TRACE it.tdlight.TelegramClient - Client 1, response received for request 6: Ok {
    }

    11:17:13.156 [TDLib thread] TRACE it.tdlight.ResponseReceiver - Received 1 events
    11:17:13.156 [TDLib thread] TRACE it.tdlight.TelegramClient - Client 1, response received for request 0: UpdateConnectionState {
    state = ConnectionStateConnecting {
    }
    }

    11:17:13.725 [TDLib thread] TRACE it.tdlight.ResponseReceiver - Received 0 events
    11:17:13.726 [TDLib thread] TRACE it.tdlight.ResponseReceiver - Received 1 events
    11:17:13.726 [TDLib thread] TRACE it.tdlight.TelegramClient - Client 1, response received for request 0: UpdateAuthorizationState {
    authorizationState = AuthorizationStateClosed {
    }
    }

    11:17:13.728 [TDLib thread] DEBUG it.tdlight.TelegramClient - Registering new client 2
    11:17:13.728 [TDLib thread] INFO  it.tdlight.TelegramClient - Registered new client 2
    11:17:13.728 [TDLib thread] DEBUG it.tdlight.TelegramClient - Sending dummy startup request as client 2
    11:17:13.728 [TDLib thread] DEBUG it.tdlight.ResponseReceiver - ResponseReceiver is now running
    11:17:13.729 [TDLib thread] TRACE it.tdlight.TelegramClient - Trying to send async request GetOption {
    name = "version"
    }

    11:17:13.729 [TDLib thread] TRACE it.tdlight.TelegramClient - Trying to execute sync request GetTextEntities {
    text = "@telegram /test_command https://telegram.org telegram.me @gif @test"
    }

    11:17:13.730 [TDLib thread] DEBUG it.tdlight.TelegramClient - Received close
    11:17:13.730 [TDLib thread] TRACE it.tdlight.ResponseReceiver - Received 4 events
    11:17:13.731 [TDLib thread] TRACE it.tdlight.TelegramClient - Client 2, response received for request 0: UpdateOption {
    name = "version"
    value = OptionValueString {
        value = "1.8.21"
    }
    }

    11:17:13.731 [TDLib thread] DEBUG it.tdlight.TelegramClient - Client closed 1
    11:17:13.731 [TDLib thread] TRACE it.tdlight.TelegramClient - Client 2, response received for request 0: UpdateOption {
    name = "commit_hash"
    value = OptionValueString {
        value = "5902a60167146b12c8ff90d78f2bfccdfad7fb50"
    }
    }

    11:17:13.731 [TDLib thread] DEBUG it.tdlight.ClientFactoryImpl - Removing Client 1 from event handlers
    11:17:13.731 [TDLib thread] TRACE it.tdlight.TelegramClient - Client 2, response received for request 0: UpdateAuthorizationState {
    authorizationState = AuthorizationStateWaitTdlibParameters {
    }
    }

    11:17:13.731 [TDLib thread] DEBUG it.tdlight.ClientFactoryImpl - Removed Client 1 from event handlers
    11:17:13.731 [TDLib thread] TRACE it.tdlight.TelegramClient - Trying to send async request SetTdlibParameters {
    useTestDc = false
    databaseDirectory = "tdlib-session\data"
    filesDirectory = "tdlib-session\downloads"
    databaseEncryptionKey = bytes [0] { }
    useFileDatabase = true
    useChatInfoDatabase = true
    useMessageDatabase = true
    useSecretChats = false
    apiId = ***
    apiHash = "***"
    systemLanguageCode = "English"
    deviceModel = "Windows 11"
    systemVersion = "10.0"
    applicationVersion = "3.2.2+td.1.8.21"
    enableStorageOptimizer = true
    ignoreFileNames = false
    }

    11:17:13.732 [TDLib thread] TRACE it.tdlight.TelegramClient - Client 2, response received for request 1: OptionValueString {
    value = "1.8.21"
    }

    11:17:13.732 [TDLib thread] TRACE it.tdlight.TelegramClient - Client 2, request event id is not registered "1", the following response has been dropped. OptionValueString {
    value = "1.8.21"
    }

    11:17:13.732 [TDLib thread] ERROR it.tdlight.TDLight - [ 0][t 0][1702199833.733003854][Client.cpp:284]	Receive must not be called simultaneously from two different threads, but this has just happened. Call it from a fixed thread, dedicated for updates and response processing.


    Process finished with exit code -1073740791 (0xC0000409)
    
</details>


`Client 2` started receiving updates! But the app immidiately crashes with exit code `-1073740791 (0xC0000409)` when I try to `SetTdlibParameters`.

I [tried](https://github.com/yvasyliev/telegram-dynamic-profile-photo/blob/soring-boot-migration/src/main/java/com/github/yvasyliev/factories/TelegramClientFactoryBean.java) to call `clientFactory.close()` right after `AuthorizationStateClosed`, but the thread gets blocked at `clientFactory.close()`, and I haven't found a solution for it.

@cavallium I updated [my project](https://github.com/yvasyliev/telegram-dynamic-profile-photo/tree/soring-boot-migration), could you please check once?
